### DISCUSSION TITLE
Identify relationship between issues and pull requests (v3 API)

### DISCUSSION THREAD

**Discussion by halfond** [community/community#24492] [Apps, API and Webhooks]
<p>Via the GH UI, I have linked a pull request P and an issue I.  I would like to be able to identify such linkages via the v3 API.  However, when I go to repo/pull/P, repo/issues/I, and repo/issues/P, I don’t see any information in the returned data that would indicate I’ve linked P and I via the GH UI.  Is there any way to find this information via the v3 API?</p>
<p>I believe my issue is similar to the issue in <a href="https://github.community/t/get-all-issues-linked-to-a-pull-request/14653">https://github.community/t/get-all-issues-linked-to-a-pull-request/14653</a>  and <a href="https://github.community/t/get-referenced-pull-request-from-issue/14027">https://github.community/t/get-referenced-pull-request-from-issue/14027</a> but I am looking for a v3 based solution as opposed to v4.</p>
<p>Thank you.</p>

**Comment by halfond** [comment, marked as answer]
<p>In case anyone has a similar issue in the future, my workaround was as follows.</p>
<p>First I grabbed the “html_url” for the issue, retrieved the contents of the URL and used an HTML parser to grab the div with class “css-truncate my-1” and then the href of the anchor within that.  The href is a relative URL to the linked pull request.  With that, I could break up the path and get the pull number, and use that number to retrieve the pull request information via the REST API.</p>
<p>It’s brittle, but it’s working…</p>

**Reply by robinboening** [reply]
Thanks for asking the question and for your workaround. I am wondering if there is really no other way?

**Reply by marcelogturing** [reply]
That workaround only works with public repositories. What can I do when I have a private repository that I need to access using REST API with a valid token?

**Comment by xmo-odoo** [comment]
Bump, this being only available from the v4 API is quite frustrating to v3 users as the access methods are so different and the v4 API is often not better or more readable.

**Reply by marcelogturing** [reply]
What is the command on v4 API? I am struggling with it. I can find the "connected" event, but the pull request is not cited in the section.

**Reply by xmo-odoo** [reply]
[`PullRequest.closingIssuesReferences`](https://docs.github.com/en/graphql/reference/objects#pullrequest)

**Comment by KastanDay** [comment]
🎉 Clean solution here. Find linked PRs from Issues and vice-versa. Fully working Python implementation (PyGithub):

```python
def get_linked_pr_from_issue(issue: Issue) -> PullRequest | None:
  """Check if the given issue has a linked pull request.

  This function iterates over the timeline of the issue and checks if there is a 'cross-referenced' event.
  If such an event is found, it checks if the source of the event is an issue and if so, it returns the issue as a pull request.

  Usage: 
  issue: Issue = repo.get_issue(number=8)
  pr_or_none = check_if_issue_has_linked_pr(issue)

  Args:
      issue (Issue): The issue to check for a linked pull request.

  Returns:
      PullRequest: The linked pull request if it exists, None otherwise.
  """
  events_pages: PaginatedList[TimelineEvent] = issue.get_timeline()
  pg_num = 0
  while events_pages.get_page(pg_num):
    page = events_pages.get_page(pg_num)
    pg_num += 1
    for e in page:
      if str(e.event) == 'cross-referenced':
        if e.source and e.source.issue:
            return e.source.issue.as_pull_request()

def get_linked_issue_from_pr(pr: PullRequest) -> Issue | None:
  """Check if the given pull request has a linked issue.

  This function iterates over the timeline of the pull request and checks if there is a 'cross-referenced' event.
  If such an event is found, it checks if the source of the event is a pull request and if so, it returns the pull request as an issue.

  Usage: 
  pr: PullRequest = repo.get_pull(number=8)
  issue_or_none = check_if_pr_has_linked_issue(pr)

  Args:
      pr (PullRequest): The pull request to check for a linked issue.

  Returns:
      Issue: The linked issue if it exists, None otherwise.
  """
  events_pages: PaginatedList[TimelineEvent] = pr.as_issue().get_timeline()
  pg_num = 0
  while events_pages.get_page(pg_num):
    page = events_pages.get_page(pg_num)
    pg_num += 1
    for e in page:
      if str(e.event) == 'cross-referenced':
        if e.source and e.source.issue:
            return e.source.issue

```

**Reply by Jericho** [reply]
@KastanDay my understanding is that `cross-referenced` event is triggered when someone adds a comment on another issue/PR that mentions the issue at hand. For example, a comment similar to: `See #xxx for more details` would trigger the cross-reference event. However it does not indicate that an issue was linked to a PR. 

Just to make sure we are on the same page: when I say that a PR and an issue are "linked", what I mean is that an association between the two has been created like in this example:
![image](https://github.com/community/community/assets/112710/d94b83fa-9b11-4ce2-a3ba-991539a1c2bb)


My experience is that `connected` is the event type that indicates an issue and a PR were linked. For testing purposes, I created an issue and I linked it to a PR. When I retrieve the timeline for this event by issuing `GET https://api.github.com/repos/jericho/_testing/issues/108/timeline?per_page=100&page=1`, this is the JSON I get:
![image](https://github.com/community/community/assets/112710/13f491c3-6c66-405a-b1d5-975af7477b7a)

My current struggle is that I am unable to figure out how to find the PR this issue was linked to based on the `connected` event. All this to say that, I may have to resort to parsing the HTML as @halfond suggests. It should be noted that in the three years since @halfond posted his suggestion, the CSS class of the element containing the desired anchor has changed as can be seen here:
![image](https://github.com/community/community/assets/112710/6bce8652-b1ec-4f36-97ba-f6856c597da6)
